### PR TITLE
podman unshare: add --rootless-cni to join the ns

### PR DESCRIPTION
**Pull request by Luap99**
Add a new --rootless-cni option to podman unshare to also join the
rootless-cni network namespace. This is useful if you want to connect
to a rootless container via IP address. This is only possible from the
rootless-cni namespace and not from the host namespace. This option also
helps to debug problems in the rootless-cni namespace.

### CODE REVIEW

File: cmd/podman/system/unshare.go [hunk 1]
```diff
@@ -33,6 +34,7 @@ func init() {
 	})
 	flags := unshareCommand.Flags()
 	flags.SetInterspersed(false)
+	flags.BoolVar(&unshareOptions.RootlessCNI, "rootless-cni", false, "Join the rootless-cni namespace")
```
Comment (mheon) [inline cmd/podman/system/unshare.go:37]: Suggest rewording description to "Join the rootless network namespace used for CNI networking"

File: pkg/domain/infra/abi/system.go [hunk 2]
```diff
@@ -390,13 +390,25 @@ func unshareEnv(graphroot, runroot string) []string {
...
+		cmd.Stderr = os.Stderr
+		return cmd.Run()
+	}
+
+	if options.RootlessCNI {
+		rootlesscni, err := ic.Libpod.GetRootlessCNINetNs(true)
+		if err != nil {
+			return err
+		}
+		defer rootlesscni.Cleanup(ic.Libpod)
```
Comment (mheon) [inline pkg/domain/infra/abi/system.go:408]: Is there a risk that we wipe out the rootless CNI namespace if we have two unshare sessions both joined to it, and one exits before the other?

Comment (Luap99) [reply pkg/domain/infra/abi/system.go:408]: Yes that is problematic. What would be the proper way to fix this? Cleanup currently checks that there are no running containers with bridge networking and that `var/lib/cni/results` is empty. Otherwise cleanup is skipped.

Comment (mheon) [reply pkg/domain/infra/abi/system.go:408]: Hm. We could try reference counting, but I'm concerned we could get into a state where we never clean up if a `podman unshare` gets SIGKILL'd before it can indicate it's done.

The easiest way is probably to guarantee there's only 1 `podman unshare` shell with `--rootless-cni` active at a time using a lockfile. I don't think we'll run into cases where more than 1 session is needed very often?

Comment (Luap99) [reply pkg/domain/infra/abi/system.go:408]: Yeah I think we should ensure that we always do cleanup. Otherwise we will not only leak the namespace but also the slirp process.

Only allowing one session feels a little odd. The only thing that will happen to the other unshare session is that we lose network connectivity because we kill the slirp process. The namespace stays alive until the process exits only the bind mount is removed bind cleanup.  Also as long as at least one container with CNI networking is running we can safely use as many sessions as we want.

I think we can live with this problem.

Comment (mheon) [reply pkg/domain/infra/abi/system.go:408]: OK, that ought to be fine, then.

File: cmd/podman/system/unshare.go (hunk 1 above)
Comment (Luap99) [reply cmd/podman/system/unshare.go]: Done

**Comment by mheon** [comment]
Compose test failing, potential flake? Restarted.

LGTM

**Comment by Luap99** [comment]
> Compose test failing, potential flake? Restarted.

Yes I have to fix this ASAP

File: pkg/domain/infra/abi/system.go (hunk 2 above)
Comment (giuseppe) [reply pkg/domain/infra/abi/system.go:408]: could we just join the network namespace without creating it?

That would also solve the issue that we don't have to cleanup on exit?

Comment (Luap99) [reply pkg/domain/infra/abi/system.go:408]: We could do this but it doesn't solve the problem. We could skip the cleanup here but if all running cni containers are stopped we will perform the cleanup and then all running unshare sessions would lose network connectivity.
Also the reason I choose to create the ns here is because this should open up more use cases. I thought about using this to create the ns and do some setup stuff, e.g. create a interface used by cni.

**Comment by mheon** [comment]
Needs a rebase

**Comment by Luap99** [comment]
Updated. I added a note to the man page that using more than one session can have unexpected results. I also added a extra check to prevent a possible panic.

**Comment by TomSweeneyRedHat** [comment]
LGTM
and happy green test buttons.

File: pkg/domain/infra/abi/system.go (hunk 2 above)
Comment (giuseppe) [reply pkg/domain/infra/abi/system.go:408]: could we split the two cases, perhaps `--rootless-cni=create|join`?  I think having an option to just join the network namespace, and lose connectivity if there are no other containers running is still a valid use case.

Comment (Luap99) [reply pkg/domain/infra/abi/system.go:408]: There is only one rootless cni namespace per user. We will always join if it exists.
